src/Auth/AuthMiddleware.php

```diff
@@ -0,0 +1,70 @@
+<?php
+namespace Yiisoft\Yii\Web\Auth;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Strings\StringHelper;
+
+final class AuthMiddleware implements MiddlewareInterface
+{
+    private const REQUEST_NAME = 'auth_user';
+
+    private $requestName = self::REQUEST_NAME;
+    private $responseFactory;
+    private $authenticator;
+    private $optional = [];
+
+    public function __construct(ResponseFactoryInterface $responseFactory, AuthInterface $authenticator)
+    {
+        $this->responseFactory = $responseFactory;
+        $this->authenticator = $authenticator;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        if ($this->isOptional($request)) {
+            return $handler->handle($request);
```

Comment:
Is it correct behavior that when authentication is optional we are not trying to authenticate user at all?

Reply:
You are right, need to improve this

Reply:
Let me know if you have a better suggestion